src/pyk/cli/args.py

```diff
@@ -1,155 +1,732 @@
+from pyk.kore.rpc import ExecuteResult, StopReason
+from pyk.kore.syntax import Pattern, kore_term
+from pyk.ktool.kprint import KPrint
+from pyk.ktool.kprove import KProve
+from pyk.prelude.k import GENERATED_TOP_CELL
+from pyk.prelude.ml import is_top, mlAnd, mlOr
+from pyk.utils import _LOGGER, ensure_dir_path
 
-from ..utils import ensure_dir_path
 from .utils import bug_report_arg, dir_path, file_path
 
 if TYPE_CHECKING:
+    from argparse import _SubParsersAction
+    from collections.abc import Iterable
     from typing import TypeVar
 
+    from ..utils import BugReport
+
     T = TypeVar('T')
 
 
-class KCLIArgs:
-    @cached_property
-    def logging_args(self) -> ArgumentParser:
-        args = ArgumentParser(add_help=False)
-        args.add_argument('--verbose', '-v', default=False, action='store_true', help='Verbose output.')
-        args.add_argument('--debug', default=False, action='store_true', help='Debug output.')
-        return args
-
-    @cached_property
-    def parallel_args(self) -> ArgumentParser:
-        args = ArgumentParser(add_help=False)
-        args.add_argument('--workers', '-j', default=1, type=int, help='Number of processes to run in parallel.')
-        return args
-
-    @cached_property
-    def bug_report_args(self) -> ArgumentParser:
-        args = ArgumentParser(add_help=False)
-        args.add_argument(
-            '--bug-report',
-            type=bug_report_arg,
-            help='Generate bug report with given name',
+class PrintInput(Enum):
+    KORE_JSON = 'kore-json'
+    KAST_JSON = 'kast-json'
+
+
+class CLI:
+    commands: list[type[Command]]
+
+    # Input a list of all Command types to be used
+    def __init__(self, commands: Iterable[type[Command]]):
+        self.commands = list(commands)
+
+    # Return an instance of the correct Options subclass by matching its name with the requested command
+    def generate_command(self, args: dict[str, Any]) -> Command:
+        command = args['command'].lower()
+        for cmd_type in self.commands:
+            if cmd_type.name() == command:
+                return cmd_type(args)
+        raise ValueError(f'Unrecognized command: {command}')
+
+    # Generate the parsers for all commands
+    def add_parsers(self, base: _SubParsersAction) -> _SubParsersAction:
+        for cmd_type in self.commands:
+            base = cmd_type.parser(base)
+        return base
+
+    def create_argument_parser(self) -> ArgumentParser:
+        pyk_args = ArgumentParser()
+        pyk_args_command = pyk_args.add_subparsers(dest='command', required=True)
+
+        pyk_args_command = self.add_parsers(pyk_args_command)
+
+        return pyk_args
+
+
+class Options:
+    def __init__(self, args: dict[str, Any]) -> None:
+        # Get defaults from this and all superclasses that define them, preferring the most specific class
+        defaults: dict[str, Any] = {}
+        mro = type(self).mro()
+        mro.reverse()
+        for cl in mro:
```

Comment:
```suggestion
        for cl in type(self).mro().reverse():
```

Can probably make this more compact, below as well.

Reply:
The reason it's written like that is because `reverse()` only updates in place, but I've changed it to use `reversed` so it can be done in one line.